### PR TITLE
feat: support delete record

### LINKED ISSUE #20: DELETE record from GreptimeDB
delete specified row(s) from GreptimeDB

## Field in Proto file

[database.proto](https://github.com/GreptimeTeam/greptime-proto/blob/main/proto/greptime/v1/database.proto#L43)

## Java Example

[quick start](https://github.com/GreptimeTeam/greptimedb-ingester-java/blob/main/ingester-example/src/main/java/io/greptime/LowLevelApiWriteQuickStart.java#L88-L95), but ingester-go should keep go convention

### PR DESCRIPTION
**Pull request by JetSquirrel**
## What's changed and what's your intention?
support delete specified row(s) from GreptimeDB

## Checklist

- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link 

This closes #20

### CODE REVIEW

File: table/types/types.go [hunk 1]
```diff
@@ -247,3 +247,13 @@ func ConvertType(type_ ColumnType) (gpb.ColumnDataType, error) {
 	}
 
 }
+
+// WriteOp is the type of write operation
+// current supported [Insert, Delete]
+// TODO: [Update]
+type WriteOp uint
```
Comment (yuanbohan) [inline table/types/types.go:254]: Is it ok to get rid of the WriteOp concept, and directly define Write and Delete functions? Which means, the `Build(WriteOp)` method of `Request`, has to be renamed to `ToDeleteRequest`, or `ToInsertRequest` or other proper names.

**Comment by yuanbohan** [comment]
Thanks for your contribution, and good documentation, and unit test. Excellent job :smile:

close #20

File: table/types/types.go (hunk 1 above)
Comment (JetSquirrel) [reply table/types/types.go:254]: Thanks @yuanbohan review and advice.I still need some help for figure out it.
 
do you mean change the `Request` to this?
```golang
interface Request{
    ...
}

type ToDeleteRequest struct {
   ...
}

type ToInsertRequest struct {
   ...
}

```

But if do this, we need implement twice the functions for `ToDeleteRequest` and `ToInsertRequest`.  Cloud you told me the benefits in this implements?

Or we can add `WriteOp` parameter in `Request`, do you think  is good method?
```golang

type Request struct {
	header *header.Header
	tables []*table.Table
        writeOp types.WriteOp
}

func write(...,writeOp,...) ... {
	...
	request_, err := request.New(header_, writeOp,tables...).Build()
        ...
}
```

Thanks for your time to help me improve code.

Comment (yuanbohan) [reply table/types/types.go:254]: I think the latter one is a better option :+1:

BTW, WriteOp is somehow confusing for delete is not a write operation, maybe direct `Operation`? or some other good names.

Comment (JetSquirrel) [reply table/types/types.go]: Done, please re-check the code, thank you.

File: client.go [hunk 2]
```diff
@@ -47,6 +48,17 @@ func NewClient(cfg *Config) (*Client, error) {
 	return &Client{cfg: cfg, client: client}, nil
 }
 
+// write is create the request and send it to GreptimeDB.
```
Comment (yuanbohan) [inline client.go:51]: write should be changed to submit. BTW, submit is a good name :+1:

File: client.go [hunk 3]
```diff
@@ -115,7 +139,35 @@ func (c *Client) WriteObject(ctx context.Context, obj any) (*gpb.GreptimeRespons
 		return nil, err
 	}
 
-	return c.Write(ctx, tbl)
+	return c.submit(ctx, types.INSERT, tbl)
+}
+
+// deleteObject is like [Delete] to delete the data in GreptimeDB, but schema is defined in the struct tag.
```
Comment (yuanbohan) [inline client.go:145]: deleteObject should be changed to DeleteObject

File: client.go [hunk 4]
```diff
@@ -115,7 +139,35 @@ func (c *Client) WriteObject(ctx context.Context, obj any) (*gpb.GreptimeRespons
...
+func (c *Client) DeleteObject(ctx context.Context, obj any) (*gpb.GreptimeResponse, error) {
+	tbl, err := schema.Parse(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.submit(ctx, types.DELETE, tbl)
+}
+
+func (c *Client) streamWrite(ctx context.Context, operation types.Operation, tables ...*table.Table) error {
```
Comment (yuanbohan) [inline client.go:156]: what about `streamSubmit` ?

File: client.go [hunk 5]
```diff
@@ -193,7 +249,17 @@ func (c *Client) StreamWriteObject(ctx context.Context, body any) error {
 	if err != nil {
 		return err
 	}
-	return c.StreamWrite(ctx, tbl)
+	return c.streamWrite(ctx, types.INSERT, tbl)
+}
+
+// StreamDeleteObject is like [StreamDelete] to send the data into GreptimeDB, but schema is defined in the struct tag.
```
Comment (yuanbohan) [inline client.go:255]: In delete scenario, **send the data into GreptimeDB** should be updated to **delete the data from GreptimeDB**

File: client_test.go [hunk 6]
```diff
@@ -189,7 +189,7 @@ func newMysql() *Mysql {
 
 func init() {
 	repo := "greptime/greptimedb"
-	tag := "v0.6.0"
+	tag := "v0.7.0"
```
Comment (yuanbohan) [inline client_test.go:192]: :+1: 
thanks for the version update

File: table/types/types.go [hunk 7]
```diff
@@ -247,3 +247,13 @@ func ConvertType(type_ ColumnType) (gpb.ColumnDataType, error) {
 	}
 
 }
+
+// WriteOp is the type of write operation
```
Comment (yuanbohan) [inline table/types/types.go:251]: `WriteOp` should be updated to `Operation`

**Review by yuanbohan — CHANGES_REQUESTED** [review]
LGTM, thanks for your commitment

**Comment by JetSquirrel** [comment]
Thank you @yuanbohan point out the mistack, I forgate change the code comments, my bad.
I already fix the mistacks.

**Review by yuanbohan — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**